Add FieldML I/O test for unit cube

rchristie opened the pull request:
As part of issue #24, want to ensure the existing code works and does not get broken by future changes.

I intend to make the FieldML I/O test into a key test, however this is currently specified in the manage scripts, not delegated to Iron/Zinc.

rchristie committed:
Fix logical AND statements expecting short circuit

Unlike C/C++, Fortran does not guarantee order of execution of terms in logical expressions.

rchristie committed:
Nullify pointers before use in field interpolation

rchristie committed:
Check pointer is associated in FieldML API

rchristie committed:
Add simple test framework

rchristie committed:
Add FieldML I/O test for unit cube

rondiplomatico commented:
The key Test spec is intentionally within manage. There is no concept of standalone installation or key tests for components. Rather, manage executes a selected set of tests (especially from components) - if you want a test from fieldml source as key test, simply invoke respective test from manage by adding it to the tests of manage. (Can't remember the file right now - using mobile github app) cheers Daniel

chrispbradley commented on src/opencmiss_iron.f90:
```diff
@@ -61522,7 +61522,9 @@ SUBROUTINE cmfe_FieldMLIO_Finalise( fieldml, err )
 #ifdef USEFIELDML
 
-    CALL FIELDML_IO_FINALISE( fieldml%fieldmlInfo, err, error, *999  )
+    IF( ASSOCIATED(fieldml%fieldmlInfo) ) THEN
```
This is probably not necessary here. The coding style is that all pointers should be checked for association before they are de-referenced. As the fieldmlInfo pointer is not de-referenced it can be left for the fieldml_io_finalise routine to check the association status. However, what the problem is here is that the fieldml_info argument (first argument to fieldml_io_finalise) should be declared as a pointer in that routine. The association can then be checked in finalise and if it is associated then the finalisation can happen (the idea is that finalise can be called with a null pointer and it just does nothing). Because it is not declared as a pointer then it gets de-referenced in the subroutine call.

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,143 @@
+!> \file
```
You probably were not aware but this is not following the code style guide i.e.,

http://cmiss.bioeng.auckland.ac.nz/OpenCMISS/doc/programmer/code_style.html

e.g., CaMeL Case etc.

rchristie replied:
I was deliberately trying to make it read like gtest which has macros EXPECT_EQ, EXPECT_NEAR. At one point I had a macro that was essentially the same as in gtest, however it looked like it would get hit by fortran's line length limit, so I decided it's best to keep it in pure Fortran.
I'll change it to CamelCase unless you say otherwise.

chrispbradley replied:
That's fine if you want those macros to be upper case to match gtest. Other internal variables and the rest of the code should be according to the code style.

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,143 @@
+  USE KINDS
+  USE OpenCMISS
+  USE, INTRINSIC :: ISO_FORTRAN_ENV, only : output_unit, error_unit
+
```
I'm not sure of the intented use but this might want to go through the WriteString routines rather than directly outputing to stdout and stderr. The WriteString routines will eventually go to stdout and stderr but also handle things like output files and parallel output etc.

rchristie replied:
WriteString routines look like they're for internal use only; the subroutines all take (...,err,error,*999)? Do you intend these to be added to the public API, or am I missing the public routines?
It does make sense to be able to get test output interspersed with internal messages.

chrispbradley replied:
Sorry, my misunderstanding that this would be called from inside the library. It is going to be called from an external program so ignore.

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,143 @@
+!>
+
+MODULE IRON_TEST_FRAMEWORK
+
```
Maybe it would be better to be OpenCMISSTestFramework as it is just a Fortran version of a unit test framework e.g. Zinc may use it for it's Fortran tests ;-)

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,143 @@
+    CHARACTER(LEN=*), INTENT(IN) :: test_name !< The name of the test.
+
+    IF (LEN(current_test_name) > 0) THEN
+      WRITE(unit = error_unit, fmt = "(a)") "Called BEGIN_TEST when test in progress. Cannot nest tests. Aborting."
```
We use inline formating for the string part of a write statement. e.g. this line would would be 

WRITE(UNIT=errorUnit, '("Called BEGIN_TEST when test in progress. Cannot nest tests. Aborting.")')

Something with a parameter to the format would be

WRITE(UNIT=outputUnit,'("Begin test: ",A)') testName

etc.

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,143 @@
+  !> Check actual value is within tolerance of expected. Reports and records any error and continues.
+  SUBROUTINE EXPECT_NEAR_DP(description, expected, actual, tolerance)
+    CHARACTER(LEN=*), INTENT(IN) :: description !< Description of quantity being compared.
+    REAL(DP), INTENT(IN) :: expected !< Expected value.
```
The integers are declared as CMISSIntg but the reals are just DP? Also, we have switched to using RP or CMISSRP meaning real precision (which will be defined to be either SP or DP depending on wether or not single or double precision is required). If specific single or double precision is desired then use SP or DP explicitly.

rchristie replied:
Is there any way to make an absolute tolerance depend on the compile-time real kind?

chrispbradley replied:
@rchristie apologies, I've just read what I wrote and it probably isn't that clear! The note about RP was just really to inform you rather than saying change this to RP. These utility functions will always be either SP or DP (with the module procedure as you have done). The routines that will call this will be RP (which will be set via a #define at compile time) and so the compiler will work about which of these to routines to call via the module procedure call. The tolerance etc. will stay the same here regardless. When it is called you would have something like EXPECT_NEAR("bob",expected,actual, 1.0E-6_RP) which would then sort it all out. Also, the comment about CMISSIntg and DP isn't clear. The CMISSIntg and CMISSRP is the public i.e., outside the library, name. The internal withing the library constant name is INTG, RP, DP etc.

rchristie replied:
Thanks - I had largely figured that out. I'm going to have CMISSDP and CMISSSP variants in here (not DP/SP).
Is Iron compiled with the RP choice, or is it purely for the client to decide? I haven't seen any internal usage of RP and the public API appears to be all DP. Is this put in place just to get ready for the future?

I'm happy to change my test code calling this to use CMISSRP, however, the tolerance value for double precision can be about a million times less than that of SP. I was merely wondering whether there is a convenient way to check the precision in use at compile time and substitute a different tolerance. Perhaps the test framework can query the kind and get a relative-to-double tolerance scale factor?

rchristie replied:
It looks like intrinsic precision() or kind() functions can be used to effect this. Will see if these can be used in an expressions when initializing tolerance parameter, otherwise it can be done at run time.

chrispbradley replied:
It is for the future as I'm just adding the RP stuff (next commit). RP is set to either SP or DP when the library is compiled. The program that calls the library should use CMISSRP always. I've added a cmfe_WorkingRealPrecisionGet(workingPrecision,err) subroutine that will return either CMFE_SINGLE_REAL_TYPE or CMFE_DOUBLE_REAL_TYPE but this is not pulled in yet. The precision() or kind() routines will take a fortran variable and tell you the precision and or kind so this is not really what you want.

rchristie replied:
According to docs, intrinsic precision() takes an 'entity' i.e. more than just a variable.
The following works for me:
`REAL(CMISSRP), PARAMETER :: tolerance = 1.0E-13_RP*(10.0_RP**(precision(0.0_DP) - precision(0.0_RP)))`
Assuming this is portable with other compilers, my only concern is that the precision of double to single (15 vs. 6) gives 9 orders of magnitude difference which may be too coarse on SP (see reasoning below). Note I'm using this tolerance to compare interpolated coordinate differences.
In your introduction of RP, are you blanket replacing DP usage internally? Are there any parts such as basis evaluation and interpolation that will remain at higher precision when building with SP? (In Zinc we used double precision dot products internally, even while generally using sp.)
Do you have an issue on github for the introduction of RP?

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,143 @@
+    MODULE PROCEDURE EXPECT_NEAR_DP
+  END INTERFACE EXPECT_NEAR
+
+  CHARACTER(LEN=:), ALLOCATABLE :: current_test_name !< The name of the current test.
```
These should probably have the SAVE attribute as they will technically lose their value when the module goes out of scope.

rchristie committed:
Use and check pointer argument in fieldml io finalise

rchristie committed:
Add single precision compare, fix write formats

rchristie committed:
Use configurable real precision

rchristie commented:
I've fixed the main issues; please re-review. I need to get on with other tasks but will continually revise this code (and promptly fix any issues people find on other compilers/OSes).

rchristie committed:
Set different tolerances for single and double precision

SP accuracy is likely to be much better than relative precision of single versus double, since internal math for interpolation is likely to be done in double precision.

chrispbradley commented on src/fieldml_util_routines.f90:
```diff
@@ -200,24 +200,28 @@ SUBROUTINE FIELDML_IO_FINALISE( FIELDML_INFO, ERR, ERROR, * )
+      CALL LIST_DESTROY( FIELDML_INFO%BASIS_HANDLES, ERR, ERROR, *999 )
+      CALL LIST_DESTROY( FIELDML_INFO%BASIS_CONNECTIVITY_HANDLES, ERR, ERROR, *999 )
+      CALL LIST_DESTROY( FIELDML_INFO%BASIS_LAYOUT_HANDLES, ERR, ERROR, *999 )
+    ELSE
```
There should be a DEALLOCATE(FIELDML_INFO) here to free memory

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,158 @@
+    MODULE PROCEDURE EXPECT_NEAR_SP
+  END INTERFACE EXPECT_NEAR
+
+  CHARACTER(LEN=:), ALLOCATABLE, SAVE :: current_test_name !< The name of the current test.
```
Just be aware that allocatable characters requires gfortran v 4.8 (I think) or later. May be an issue e.g., native gfortran on hpc3 is 4.4.7.

rchristie replied:
Looks like this was implemented for scalars only in v4.7, perhaps it also needed fixes in 4.8. Okay I'll have to just allocate it myself. No worries. Will do tomorrow.

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,158 @@
+  SUBROUTINE INITIALISE_TESTS()
+    current_test_name = ""
+    test_err = 0
+  END SUBROUTINE INITIALISE_TESTS
```
Were you going to change this to CaMeL case? i.e., currentTestName and testErr?

rchristie replied:
No, I wanted this to look and feel more like gtest.

chrispbradley replied:
Yes, that was for the routine names and that is fair enough but is there any reason why the rest of the code can't follow the style guide? This file is a little bit funny in that it is technically outside the library and therefor on one hand you could make the argument that because it is on the outside it doesn't need to follow the style guide. On the other hand, I assume that the idea is that this is distributed with the rest of the library code and therefore it should try to be like the rest of the library code and follow the style guide.

rchristie committed:
Call deallocate in FieldML Info finalise

chrispbradley commented on src/fieldml_util_routines.f90:
```diff
@@ -202,23 +202,11 @@ SUBROUTINE FIELDML_IO_FINALISE( FIELDML_INFO, ERR, ERROR, * )
       CALL LIST_DESTROY( FIELDML_INFO%BASIS_LAYOUT_HANDLES, ERR, ERROR, *999 )
+      DEALLOCATE(FIELDML_INFO)
     ELSE
       CALL FlagError("FieldML Info is not associated.",ERR,ERROR,*999)
```
Sorry, I should have seen this earlier. We don't have else flag error if it is not associated in finalise routines. This is so finalise can be called on error condition from initialise. The (user callable) destroy routines (which are effectively just a wrapper for finalise) do, however, check for association.

rchristie committed:
Remove allocatable current test name

rchristie committed:
Use CamelCase internally in test framework

rchristie committed:
Remove error if not associated in finalise

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -66,53 +66,53 @@ MODULE IRON_TEST_FRAMEWORK
-  SUBROUTINE set_err()
-    test_err = 1
+  SUBROUTINE setError()
+    testErr = 1
```
Subroutine names and types are upper CaMeL case i.e., SetError. Only variable names are lower CaMeL case.

chrispbradley commented on src/fieldml_util_routines.f90:
```diff
@@ -207,8 +207,6 @@ SUBROUTINE FIELDML_IO_FINALISE( FIELDML_INFO, ERR, ERROR, * )
       CALL LIST_DESTROY( FIELDML_INFO%BASIS_CONNECTIVITY_HANDLES, ERR, ERROR, *999 )
       CALL LIST_DESTROY( FIELDML_INFO%BASIS_LAYOUT_HANDLES, ERR, ERROR, *999 )
       DEALLOCATE(FIELDML_INFO)
```
This is great that you tidied up the finalise with a pointer but I just checked and the corresponding initialise doesn't use a pointer or allocate the fieldml info variable? I'm surprised it runs? Maybe have a look in LIST_INITIALISE/LIST_FIANLISE to see how it is done? (NOTE: these routines haven't been converted to CaMeL case yet.). Note the use of the 998 label in the initialise routine to skip the finalise call to deallocate on error.

rchristie replied:
The internal FieldML Info is a pointer member of the external object. A call to public cmfe_FieldMLIO_Initialise only NULLIFYs the internal pointer; this appears to be consistent with other types.

The internal pointers themselves are allocated in the respective public CreateStart functions which are separate for Input and Output and an internal flag is set stating this 'direction'. The allocated internal pointer is passed as a non-pointer INTENT(INOUT) dummy argument to the internal initialise which fills its contents. This is an example of where it is allocated (in public cmfe_FieldML_OutputCreateObjVS):

`ALLOCATE( fieldml%fieldmlInfo, stat=err )`
`CALL FIELDML_OUTPUT_INITIALISE_INFO( mesh%mesh, location, baseName, connectivityFormat, fieldml%fieldmlInfo, &
      & err, error, *999 )`

The use of INTENT(INOUT) for a non-pointer dummy argument is probably fine for receiving already allocated pointers. However, it appears these allocations are not checked for success.

While in general test code submissions such as this should not fix internal issues that they do not test, if you agree that it is an improvement in style I'll move the allocations inside the internal CreateStart calls and test them, and obviously pass pointers as far as where allocations/deallocations occur. In all other places a pointer that's already proven as ASSOCIATED should be passed as an INTENT(INOUT) dummy argument so no further checks are needed.

Pah it doesn't look like the internal pointer is checked for being associated in most other API functions... can see if there's a tidy fix for this. I can certainly make a test which fails by crashing in this case.

chrispbradley replied:
Yes, Caton wrote these routines and he didn't follow what had been done before. The data should be allocated in the initialise routine called from CreateStart etc. All allocations should be checked for error. We try not to dereference pointers across subroutine calls. Note that pointers can also have intent attributes. The intent means the same thing. INTENT(IN) means the pointer will not be changed. INTENT(OUT) means the pointer must be assigned and INTENT(INOUT) means both.
With regard to fixing the internal issues then I didn't mean for you to fix it. I was just pointing out that it was done incorrectly so that you didn't think that was how things are done. However, once you changed the finalise routine then the initialise also had to be changed to match. We can fix up all the other code later.

chrispbradley commented on tests/FieldML_IO/cube.f90:
```diff
@@ -0,0 +1,251 @@
+
+MODULE IRON_TEST_FIELDML_CUBE
+
+  USE KINDS
```
You shouldn't need to use KINDS here. They should be acessible from the OpenCMISS and OpenCMISS_Iron modules.

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,159 @@
+MODULE IRON_TEST_FRAMEWORK
+
+  USE KINDS
+  USE OpenCMISS
```
You shouldn't need KINDS here. OpenCMISS should have them defined.

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,159 @@
+  SUBROUTINE END_TEST()
+    IF (LEN(TRIM(currentTestName)) == 0) THEN
+      WRITE(unit = error_unit, fmt = '("Called END_TEST without call to BEGIN_TEST. Aborting.")')
+      STOP 1
```
We don't really use unit= and fmt= We just use WRITE(error_unit,'(".....")')

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,159 @@
+    REAL(CMISSDP), INTENT(IN) :: tolerance !< Absolute tolerance actual value must be around expected value
+
+    IF ((actual < (expected - tolerance)).OR.(actual > (expected + tolerance))) THEN
+      WRITE(unit = output_unit, fmt = '("The value of ",a," (",g0,") is not near expected (",g0,") with tolerance (",g0,")")') &
```
The IF could just be IF(ABS(actual-expected)>tolerance) THEN

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,159 @@
+    REAL(CMISSSP), INTENT(IN) :: tolerance !< Absolute tolerance actual value must be around expected value
+
+    IF ((actual < (expected - tolerance)).OR.(actual > (expected + tolerance))) THEN
+      WRITE(unit = output_unit, fmt = '("The value of ",a," (",g0,") is not near expected (",g0,") with tolerance (",g0,")")') &
```
DItto with the previous IF comment

rchristie committed:
Fix issues from code review

rchristie committed:
Allocate FieldML info in its initialise routine

Also check allocation success and clean up partially constructed object.

rchristie committed:
Use CamelCase for cube test routines

chrispbradley commented on tests/Utilities/testframework.f90:
```diff
@@ -0,0 +1,158 @@
+  USE OpenCMISS
+  USE, INTRINSIC :: ISO_FORTRAN_ENV, only : output_unit, error_unit
+
+  IMPLICIT NONE
```
Note: ISO_FORTRAN_ENV requires gfortran v 4.5 or greater

rchristie commented:
I've fixed the allocations and CamelCase issues in the last commits.
I'll fix the issue of checking API arguments (that the FieldML Info is associated) on the next pull request, since it involves adding a new test.

rchristie committed:
Remove ISO_FORTRAN_ENV from test framework
